File: frontend/src/types/annotation.ts
export interface Annotation {
  id: string;
  documentId: string;
  userId: string;
  userName: string;
  userAvatar?: string;
  createdAt: Date;
  updatedAt?: Date;
  
  // Position in document
  target: AnnotationTarget;
  
  // Annotation content
  body: AnnotationBody;
  
  // Threading and status
  replies?: Annotation[];
  replyTo?: string;
  resolved: boolean;
  resolvedBy?: string;
  resolvedAt?: Date;
  
  // For PDF documents
  pageNumber?: number;
  
  // Permissions
  canEdit: boolean;
  canDelete: boolean;
}

export interface AnnotationTarget {
  selector: TextSelector | XPathSelector;
  source?: string;  // Document URL or ID
}

export interface TextSelector {
  type: 'TextQuoteSelector' | 'TextPositionSelector';
  exact: string;      // The exact quoted text
  prefix?: string;    // Text before the quote
  suffix?: string;    // Text after the quote
  start?: number;     // Start position
  end?: number;       // End position
}

export interface XPathSelector {
  type: 'XPathSelector';
  xpath: string;
  offset?: number;
}

export interface AnnotationBody {
  type: 'TextualBody' | 'Highlight' | 'Tag';
  value: string;
  purpose: 'commenting' | 'highlighting' | 'tagging' | 'questioning';
  color?: string;  // For highlights
  tags?: string[]; // For categorization
}

export interface CreateAnnotationDto {
  documentId: string;
  target: AnnotationTarget;
  body: AnnotationBody;
  replyTo?: string;
}

File: backend/app/schemas/annotation.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID

class AnnotationTarget(BaseModel):
    selector: Dict[str, Any]
    source: Optional[str] = None

class AnnotationBody(BaseModel):
    type: str  # 'TextualBody', 'Highlight', 'Tag'
    value: str
    purpose: str  # 'commenting', 'highlighting', 'tagging', 'questioning'
    color: Optional[str] = None
    tags: Optional[List[str]] = None

class AnnotationBase(BaseModel):
    target: AnnotationTarget
    body: AnnotationBody
    page_number: Optional[int] = None

class AnnotationCreate(AnnotationBase):
    document_id: int
    reply_to: Optional[UUID] = None

class AnnotationUpdate(BaseModel):
    body: Optional[AnnotationBody] = None
    resolved: Optional[bool] = None

class AnnotationInDBBase(AnnotationBase):
    id: UUID
    document_id: int
    user_id: int
    reply_to: Optional[UUID] = None
    thread_id: Optional[UUID] = None
    resolved: bool
    resolved_by: Optional[int] = None
    resolved_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class Annotation(AnnotationInDBBase):
    pass

class AnnotationResponse(AnnotationInDBBase):
    user_name: str
    user_avatar: Optional[str] = None
    can_edit: bool = False
    can_delete: bool = False

class AnnotationWithReplies(AnnotationResponse):
    replies: List[AnnotationResponse] = []